src/gtimelog/main.py

```diff
@@ -172,6 +174,143 @@ def make_option(long_name, short_name=None, flags=0, arg=GLib.OptionArg.NONE,
+
+        if self.gnomekeyring is None:
+            callback(username, password)
+            return
+
+        # FIXME - would be nice to make all keyring calls async, to dodge
+        # the possibility of blocking the UI. The code is all set up for
+        # that, but there's no easy way to use the keyring asynchronously
+        # from Python (as of Gnome 3.2)...
+        l = self.gnomekeyring.find_network_password_sync (
+                None,           # user
+                uri.get_host(), # domain
+                uri.get_host(), # server
+                None,           # object
+                uri.get_scheme(),# protocol
+                None,           # authtype
+                uri.get_port()) # port
+
+        result, keys = l
+        if result == self.gnomekeyring.Result.NO_MATCH:
+            # We didn't find any passwords, just continue
+            pass
+        elif result == self.gnomekeyring.Result.NO_KEYRING_DAEMON:
+            pass
+        else:
+            entry = keys[-1] # take the last key (Why?)
+            username = entry.user
+            password = entry.password
+
+        callback(username, password)
+
+    def save_to_keyring(self, uri, username, password):
+        self.gnomekeyring.set_network_password_sync (
+                None,           # keyring
+                username,       # user
+                uri.get_host(), # domain
+                uri.get_host(), # server
+                None,           # object
+                uri.get_scheme(),# protocol
+                None,           # authtype
+                uri.get_port(), # port
+                password)       # password
+
+    def ask_the_user(self, auth, uri, callback):
+        """Pops up a username/password dialog for uri"""
+        mountoperation = Gtk.MountOperation.new()
+
+        def on_reply(m, r):
+            save_to_keyring = self.gnomekeyring and (m.get_password_save() == Gio.PasswordSave.PERMANENTLY)
+
+            if r == Gio.MountOperationResult.HANDLED:
+                username = m.get_username()
+                password = m.get_password()
+
+                if username and password:
+                    if self.gnomekeyring and (m.get_password_save() == Gio.PasswordSave.PERMANENTLY):
+                        self.save_to_keyring(uri, username, password)
+                    elif m.get_password_save() == Gio.PasswordSave.FOR_SESSION:
+                        self.username = username
+                        self.password = password
+
+            else:
+                username = None
+                password = None
+
+            callback(username, password)
+
+        mountoperation.connect('reply', on_reply)
+        mountoperation.set_password_save(True)
+        mountoperation.do_ask_password(mountoperation, 'Authentication Required', '', auth.get_realm(), Gio.AskPasswordFlags.NEED_PASSWORD | Gio.AskPasswordFlags.NEED_USERNAME | (Gio.AskPasswordFlags.SAVING_SUPPORTED if self.gnomekeyring else 0))
```

Comment:
Should `Authentication Required` be translatable?

Reply:
Probably!